service/topology_coordinator.cc

```diff
@@ -991,6 +991,23 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
                 }
             };
 
+            auto check_excluded_replicas = [&] {
+                    auto tsi = get_migration_streaming_info(get_token_metadata().get_topology(), tmap.get_tablet_info(gid.tablet), trinfo);
```

Comment:
I've realized now that the fact that we're using topology from get_token_metadata() here may not be safe. for_each_tablet_transition() is yielding while iterating, which means that the topology version we use here to interpret trinfo may be different than the one from trinfo comes from, and the two may be inconsistent. For example, topology may be missing info for replicas. It could happen that this coordinator was demoted and another one took over while we're iterating here. The scenario is not very likely and any computation here in this case will be dropped because group0 update will fail if the coordinator is demoted, but violation of invariants could maybe cause some bad effects like undefined behavior or abort. To be safe, we should work with the same token metadata snapshot which we use to iterate over tablets.

Reply:
On the second thought, this runs under group0 guard, which means that token metadata cannot change locally, so we're safe.